From c00126f68254f1a16c1723fc1415ff9b14a543ea Mon Sep 17 00:00:00 2001
From: eremanikita <eremanikita@gmail.com>
Date: Sat, 22 Mar 2025 17:17:58 +0300
Subject: [PATCH 2/6] add addmob command and add json serialization

---
 20250317/1/client.py | 72 +++++++++++++++++++++++++++++++-------------
 1 file changed, 51 insertions(+), 21 deletions(-)

diff --git a/20250317/1/client.py b/20250317/1/client.py
index fb99e98..714f6dd 100644
--- a/20250317/1/client.py
+++ b/20250317/1/client.py
@@ -2,14 +2,9 @@ import readline
 import cmd
 import sys
 import socket
+import shlex
 from enum import Enum
-
-
-class Direction(Enum):
-    UP = "0 -1"
-    DOWN = "0 1"
-    RIGHT = "1 0"
-    LEFT = "-1 0"
+import json
 
 
 class Cord:
@@ -17,42 +12,76 @@ class Cord:
         self.x: int = x
         self.y: int = y
 
+    def to_dict(self):
+        return {"x": self.x, "y": self.y}
+
     def __str__(self):
         return f"({self.x}, {self.y})"
 
 
+class Direction(Enum):
+    UP = Cord(0, -1)
+    DOWN = Cord(0, 1)
+    RIGHT = Cord(1, 0)
+    LEFT = Cord(-1, 0)
+
+
+class ParserService:
+    required_keys = {"name", "hello", "hp", "coords"}
+
+    @staticmethod
+    def parse_addmob(line: str):
+        tokens = shlex.split(line)
+        params = {"name": tokens[0]}
+        i = 1
+        token_length = len(tokens)
+        while i < token_length:
+            if tokens[i] == "hello":
+                if i + 1 < token_length:
+                    params["hello"] = tokens[i + 1]
+                i += 2
+            elif tokens[i] == "hp":
+                if i + 1 < len(tokens) and tokens[i + 1].isdigit() and int(tokens[i + 1]) > 0:
+                    params["hp"] = int(tokens[i + 1])
+                i += 2
+            elif tokens[i] == "coords":
+                if i + 2 < len(tokens) and tokens[i + 1].isdigit() and tokens[i + 2].isdigit():
+                    params["coords"] = Cord(int(tokens[i + 1]), int(tokens[i + 2])).to_dict()
+                i += 3
+            else:
+                i += 1
+
+        return params if not (ParserService.required_keys - params.keys()) else None
+
+
 class MUDGame(cmd.Cmd):
     prompt = "command>> "
 
     def do_up(self, args):
         """Move the player up"""
-        s.sendall(f"move {Direction.UP}".encode())
+        s.sendall((json.dumps({"command": "move", "params": Direction.UP.value.to_dict()}) + "\n").encode())
+        print(s.recv(1024).rstrip().decode())
 
     def do_down(self, args):
         """Move the player down"""
-        s.sendall(f"move {Direction.DOWN}".encode())
+        s.sendall((json.dumps({"command": "move", "params": Direction.DOWN.value.to_dict()}) + "\n").encode())
 
     def do_left(self, args):
         """Move the player left"""
-        s.sendall(f"move {Direction.LEFT}".encode())
+        s.sendall((json.dumps({"command": "move", "params": Direction.LEFT.value.to_dict()}) + "\n").encode())
 
     def do_right(self, args):
         """Move the player right"""
-        s.sendall(f"move {Direction.RIGHT}".encode())
+        s.sendall((json.dumps({"command": "move", "params": Direction.RIGHT.value.to_dict()}) + "\n").encode())
 
     def do_addmob(self, args):
         """Add a mob to the field
         addmob <monster_name> hello <hello_string> hp <hitpoints> coords <x> <y>"""
-        pass
-
-    def do_attack(self, args):
-        """Attack the specific monster in the same cell if it is on it
-        weapons:
-            sword: -10 hp
-            spear: -15 hp
-            axe: -20 hp
-        nothing if there is no monster in the same cell or the name is incorrect"""
-        pass
+        params = ParserService.parse_addmob(args)
+        if params:
+            s.sendall((json.dumps({"command": "addmob", "params": params}) + "\n").encode())
+        else:
+            print("Invalid command")
 
     def default(self, line):
         print("Invalid command")
@@ -63,5 +92,6 @@ if __name__ == "__main__":
     host = "localhost" if len(sys.argv) < 2 else sys.argv[1]
     port = 1337 if len(sys.argv) < 3 else int(sys.argv[2])
     with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
         s.connect((host, port))
         MUDGame().cmdloop()
-- 
2.39.5 (Apple Git-154)

